swap/int256/int256.go

```diff
@@ -0,0 +1,161 @@
+	v := new(Int256)
+	v.value = new(big.Int).Set(u.value)
+	return v
+}
+
+// Value returns the underlying private value for a Int256 struct
+func (u *Int256) Value() *big.Int {
+	return new(big.Int).Set(u.value)
+}
+
+// set assigns the underlying value of the given Int256 param to u, and returns the modified receiver struct
+// returns an error when the value cannot be correctly set
+func (u *Int256) set(value *big.Int) (*Int256, error) {
+	if err := checkInt256Bounds(value); err != nil {
+		return nil, err
+	}
+	if u.value == nil {
+		u.value = new(big.Int)
+	}
+	u.value.Set(value)
+	return u, nil
+}
+
+// checkInt256Bounds returns an error when the given value falls outside of the signed 256-bit integer range
+// returns nil otherwise
+func checkInt256Bounds(value *big.Int) error {
+	if value.Cmp(maxInt256) == 1 {
+		return fmt.Errorf("cannot set Int256 to %v as it overflows max value of %v", value, maxInt256)
+	}
+	if value.Cmp(minInt256) == -1 {
+		return fmt.Errorf("cannot set Int256 to %v as it underflows min value of %v", value, minInt256)
+	}
+	return nil
+}
+
+// Add sets u to augend + addend and returns u as the sum
+// returns an error when the value cannot be correctly set
+func (u *Int256) Add(augend, addend *Int256) (*Int256, error) {
+	sum := new(big.Int).Add(augend.value, addend.value)
+	return u.set(sum)
+}
+
+// Sub sets u to minuend - subtrahend and returns u as the difference
+// returns an error when the value cannot be correctly set
+func (u *Int256) Sub(minuend, subtrahend *Int256) (*Int256, error) {
+	difference := new(big.Int).Sub(minuend.value, subtrahend.value)
+	return u.set(difference)
+}
+
+// Mul sets u to multiplicand * multiplier and returns u as the product
+// returns an error when the value cannot be correctly set
+func (u *Int256) Mul(multiplicand, multiplier *Int256) (*Int256, error) {
+	product := new(big.Int).Mul(multiplicand.value, multiplier.value)
+	return u.set(product)
+}
+
+// cmp calls the underlying Cmp method for the big.Int stored in a Int256 struct as its value field
+func (u *Int256) cmp(v *Int256) int {
+	return u.value.Cmp(v.value)
+}
+
+// Equals returns true if the two Int256 structs have the same underlying values, false otherwise
+func (u *Int256) Equals(v *Int256) bool {
+	return u.cmp(v) == 0
+}
+
+// String returns the string representation for Int256 structs
+func (u *Int256) String() string {
+	return u.value.String()
+}
+
+// MarshalJSON implements the json.Marshaler interface
+// it specifies how to marshal a Int256 struct so that it can be written to disk
+func (u *Int256) MarshalJSON() ([]byte, error) {
+	return []byte(u.value.String()), nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface
+// it specifies how to unmarshal a Int256 struct so that it can be reconstructed from disk
+func (u *Int256) UnmarshalJSON(b []byte) error {
+	if string(b) == "null" {
+		return nil
+	}
+
+	var value big.Int
+	_, ok := (&value).SetString(string(b), 10)
```

Comment:
I am sorry if this was already discussed, but could a new big.Int can be avoided by calling SetString on u.value (with checkInt256Bounds)?

Reply:
i think that at this point `u.value` is nil that's why `set` is called (and in turn creates the value instance).
@mortelli I think it is ok to have the entire value reconstruction logic here explicit, then skip the `set` call as @janos suggests. You can have a look at other serialization/deserialization functions in swarm - we usually manipulate the structs directly from the (de)serialization methods.

Reply:
Thanks @acud. I have to say that this code works and that nil argument is valid. So, it is fine for me in this form also, and if needed, can be changed later.

Reply:
not entirely sure what my alternative is here. both `set` and `checkInt256Bounds` require `big.Int` pointers as arguments, so if i were to set call `SetString` directly on `value`, wouldn't i have to change it back if something went wrong when checking it afterwards? 

or would it be ok to just return an error after having set the value anyways?